onnxruntime/core/session/custom_ops.cc

```diff
@@ -395,17 +398,272 @@ struct CustomOpKernel : OpKernel {
+    }
+    std::string input_name = "Input" + std::to_string(i);
+    schema.Input(i, input_name, "", input_name, option, is_homogeneous, min_arity);
+    schema.TypeConstraint(input_name, DataTypeImpl::ToString(DataTypeImpl::AllTensorTypes()), "all types");
+  }
+
+  for (size_t i = 0; i < output_count; i++) {
+    onnx::OpSchema::FormalParameterOption option = onnx::OpSchema::FormalParameterOption::Single;
+    bool is_homogeneous = true;
+    int min_arity = 1;
+
+    // The OrtCustomOp interface did not support the methods to query input/output characteristics before
+    // ORT API version 8. So, query the relevant methods ONLY from API version 8 onwards.
+    if (op->version >= min_ort_version_with_optional_io_support) {
+      const auto characteristic = op->GetOutputCharacteristic(op, i);
+
+      // Support for optional and variadic inputs/output was added in versions 8 and 14, respectively.
+      if (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL) {
+        option = onnx::OpSchema::FormalParameterOption::Optional;
+      } else if ((op->version >= min_ort_version_with_variadic_io_support) &&
+                 (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC)) {
+        ORT_ENFORCE(i == output_count - 1, "Only the last output to a custom op may be marked variadic.");
+        option = onnx::OpSchema::FormalParameterOption::Variadic;
+        min_arity = op->GetVariadicOutputMinArity(op);
+        is_homogeneous = static_cast<bool>(op->GetVariadicOutputHomogeneity(op));
+      }
+    }
+    const auto type = op->GetOutputType(op, i);
+    if (ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED == type) {
+      if (op->GetOutputCharacteristic(op, i) == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_REQUIRED) {
+        ORT_ENFORCE(1 == undefined,
+                    "There must be one (and only one) dynamic typed input to the custom op. "
+                    "Its type info at runtime will be used to infer the type info of this dynamic typed output "
+                    "which is required for the success of the model loading step. "
+                    "More than one dynamic typed inputs are currently not supported as differing types at runtime means the output type "
+                    "cannot be inferred without which model loading cannot proceed.");
+      }
+    }
+    std::string output_name = "Output" + std::to_string(i);
+    schema.Output(i, output_name, "", output_name, option, is_homogeneous, min_arity);
+    schema.TypeConstraint(output_name, DataTypeImpl::ToString(DataTypeImpl::AllTensorTypes()), "all types");
+  }
+  schema.SetDomain(domain);
+  schema.SinceVersion(1);
+  schema.AllowUncheckedAttributes();
+  return schema;
+}
+
+Status IsCompatible(const ONNX_NAMESPACE::OpSchema& schema, const OrtCustomOp* op) {
+  const size_t input_count = op->GetInputTypeCount(op);
+  const size_t output_count = op->GetOutputTypeCount(op);
+
+  // check inputs
```

Comment:
should we check other properties, like characteristic?

Reply:
Characteristic has been included.
Beyond that we also check arity and homogeneity.